olive/resource_path.py

```diff
@@ -0,0 +1,379 @@
+        model_path = Path(model.path)
+
+        # path to save the resource to
+        if name:
+            new_path_name = Path(name).with_suffix(model_path.suffix).name
+        else:
+            new_path_name = model_path.name
+        new_path = dir_path / new_path_name
+        _overwrite_helper(new_path, overwrite)
+
+        # download the resource to the new path
+        logger.debug(f"Downloading model {self.name} version {self.version} to {new_path}.")
+        from azure.core.exceptions import ServiceResponseError
+
+        with tempfile.TemporaryDirectory(dir=dir_path, prefix="olive_tmp") as temp_dir:
+            temp_dir = Path(temp_dir)
+            retry_func(
+                ml_client.models.download,
+                [self.name],
+                {"version": self.version, "download_path": temp_dir},
+                max_tries=3,
+                delay=5,
+                exceptions=ServiceResponseError,
+            )
+            new_path.parent.mkdir(parents=True, exist_ok=True)
+            shutil.move(temp_dir / self.name / model_path.name, new_path)
+
+        return str(new_path)
+
+
+class AzureMLDatastoreConfig(ResourceConfig):
+    """Config for an AzureML datastore resource."""
+
+    _type = ResourceType.AzureMLDatastore
+    azureml_client: AzureMLClientConfig = Field(..., description="AzureML client config.")
+    datastore_name: str = Field(..., description="Name of the datastore.")
+    relative_path: str = Field(..., description="Relative path to the resource from the datastore root.")
+
+    def get_path(self) -> str:
+        workspace_config = self.azureml_client.get_workspace_config()
+        return (
+            f"{_get_azureml_resource_prefix(workspace_config)}"
+            f"/datastores/{self.datastore_name}/paths/{self.relative_path}"
+        )
+
+    def save_to_dir(self, dir_path: Union[Path, str], name: str = None, overwrite: bool = False) -> str:
+        # there is no direct way to download a file from a datastore
+        # so we will use a workaround to download the file by creating a aml model
+        # that references the file and downloading the model
+        from azure.ai.ml.constants import AssetTypes
+        from azure.ai.ml.entities import Model
+        from azure.core.exceptions import ServiceResponseError
+
+        # azureml client
+        ml_client = self.azureml_client.create_client()
+
+        # create aml model
+        logger.debug(f"Creating aml model for datastore {self.datastore_name} path {self.relative_path}.")
+        aml_model = retry_func(
+            ml_client.models.create_or_update,
+            [
+                Model(
+                    path=self.get_path(),
+                    name="olive-backend-model",
+                    description="Model created by Olive backend. Ignore this model.",
+                    type=AssetTypes.CUSTOM_MODEL,
+                )
+            ],
+            max_tries=3,
```

Comment:
How about to extract the `max_tries` or `delay` to default configs? If there is a few network issue, use can adjust the numbers.

Reply:
Done. Exposed in `AzureMLClientConfig`